packages/clay-date-picker/src/index.tsx

```diff
@@ -281,13 +292,65 @@ const ClayDatePicker: React.FunctionComponent<IProps> = React.forwardRef<
 			}
 		};
 
+		/**
+		 * Holds the logic for how to behave when clicking on a date,
+		 * on the DatePicker using date ranges
+		 */
+		const handleDateRangeClicked = (
+			date: Date,
+			daysSelected: TDaysSelected
+		) => {
+			const [startDate, endDate] = daysSelected;
+
+			if (date < startDate) {
+				setDaysSelected([date, endDate]);
+
+				onValueChange(
+					fromRangeToString(daysSelected, dateFormat),
+					'click'
+				);
+
+				return;
+			}
+
+			setDaysSelected([startDate, date]);
+
+			onValueChange(
+				fromRangeToString([startDate, date], dateFormat),
+				'click'
+			);
+		};
+
 		/**
 		 * Handles the click on element of the day
 		 */
 		const handleDayClicked = (date: Date) => {
-			setDaySelected(date);
+			if (range) {
+				handleDateRangeClicked(date, daysSelected);
+
+				return;
+			}
 
-			onValueChange(date, 'click');
+			setDaysSelected([date, date]);
+
+			const dateFormatted = formatDate(date, dateFormat);
+
+			if (time) {
+				const dateString = formatDate(
+					parseDate(
+						`${dateFormatted} ${currentTime}`,
+						`${dateFormat} ${TIME_FORMAT}`,
+						date
+					),
+					`${dateFormat} ${TIME_FORMAT}`
+				);
+
+				onValueChange(dateString, 'click');
+
+				return;
+			}
+
+			onValueChange(dateFormatted, 'click');
```

Comment:
Why not keep it all together instead of separating it into two functions, just thinking about simplifying it a little bit here, maybe something like this:

```js
/**
 * Handles the click on element of the day
 */
const handleDayClicked = (date: Date) => {
	const [startDate, endDate] = daysSelected;

	let newDaysSelected ;
	let daysSelectedToString;

	if (range) {
		newDaysSelected = date < startDate ? [date, endDate] : [startDate, date];

		daysSelectedToString = fromRangeToString(newDaysSelected, dateFormat);
	} else {
		// Ideally, just set a date: [date]
		newDaysSelected = [date, date];

		daysSelectedToString = formatDate(date, dateFormat);

		if (time) {
			daysSelectedToString = formatDate(
				parseDate(
					`${daysSelectedToString} ${currentTime}`,
					`${dateFormat} ${TIME_FORMAT}`,
					date
				),
				`${dateFormat} ${TIME_FORMAT}`
			);
		}
	}

	setDaysSelected(newDaysSelected);
	onValueChange(daysSelectedToString, 'click');
}
```

You can avoid repeatedly calling `onValueChange` and `setDaysSelected` and you can slightly improve the readability of the use cases.

The bad thing about maintaining it as in PR is that it is difficult to read the use cases that exist, seeing many calls of `onValueChange` or `setDaysSelected` is like creating several blocks of edge cases and it is difficult to make one sequential reading of what is happening. 

Separating into another function can help as you did but it will just repeat more code and more lines and maybe complicate it a little more because it will not be reused elsewhere.